core/src/main/scala-3/spire/syntax/macros/cforMacros.scala

```diff
@@ -15,157 +15,117 @@
-    var index = $fromExpr
-    val limit = $untilExpr
-    while (index < limit) {
-      ${ Expr.betaReduce(body) }(index)
-      index += $stride
-    }
-  }
+end RangeForImpl
 
-  def strideUpTo(fromExpr: Expr[Int], untilExpr: Expr[Int], stride: Expr[Int]): Expr[Unit] = '{
-    var index = $fromExpr
-    val end = $untilExpr
-    while index <= end do
-      ${ Expr.betaReduce(body) }(index)
-      index += $stride
-  }
+private def letFunc[A, B, C](using Quotes)(name: String, rhs: Expr[A => B])(in: Expr[A => B] => Expr[C]): Expr[C] =
+  import quotes.reflect.*
 
-  def strideDownTo(fromExpr: Expr[Int], untilExpr: Expr[Int], stride: Expr[Int]): Expr[Unit] = '{
-    var index = $fromExpr
-    val end = $untilExpr
-    while index >= end do
-      ${ Expr.betaReduce(body) }(index)
-      index -= $stride
-  }
+  extension (t: Term) def unsafeAsExpr[A] = t.asExpr.asInstanceOf[Expr[A]] // cast without `quoted.Type[A]`
 
-  def strideDownUntil(fromExpr: Expr[Int], untilExpr: Expr[Int], stride: Expr[Int]): Expr[Unit] = '{
-    var index = $fromExpr
-    val limit = $untilExpr
-    while index > limit do
-      ${ Expr.betaReduce(body) }(index)
-      index -= $stride
-  }
+  def isFunctionLiteral[A, B](f: Expr[A => B]): Boolean =
+    f.asTerm.underlyingArgument match
+      case Block(List(ddef), Closure(ref, _)) => ref.symbol == ddef.symbol
+      case _                                  => false
 
-  r match
-    case '{ ($i: Int) until $j } => strideUpUntil(i, j, Expr(1))
-    case '{ ($i: Int) to $j }    => strideUpTo(i, j, Expr(1))
-    case '{ ($i: Int) until $j by $step } =>
-      step.asTerm match {
-        case Literal(IntConstant(k)) if k > 0  => strideUpUntil(i, j, Expr(k))
-        case Literal(IntConstant(k)) if k < 0  => strideDownUntil(i, j, Expr(-k))
-        case Literal(IntConstant(k)) if k == 0 => report.error("zero stride", step); '{}
-        case _ =>
-          report.warning(s"defaulting to foreach, can not optimise non-constant step", step)
-          '{ val b = $body; $r.foreach(b) }
-      }
-    case '{ ($i: Int) to $j by $step } =>
-      step.asTerm match {
-        case Literal(IntConstant(k)) if k > 0  => strideUpTo(i, j, Expr(k))
-        case Literal(IntConstant(k)) if k < 0  => strideDownTo(i, j, Expr(-k))
-        case Literal(IntConstant(k)) if k == 0 => report.error("zero stride", step); '{}
-        case _ =>
-          report.warning(s"defaulting to foreach, can not optimise non-constant step", step)
-          '{ val b = $body; $r.foreach(b) }
-      }
-    case _ =>
-      report.warning(s"defaulting to foreach, can not optimise range expression", r)
-      '{ val b = $body; $r.foreach(b) }
+  def let[A, B](name: String, f: Expr[A])(in: Expr[A] => Expr[B]): Expr[B] =
+    ValDef.let(Symbol.spliceOwner, name, f.asTerm)(ref => in(ref.unsafeAsExpr[A]).asTerm).unsafeAsExpr[B]
 
-end fastForRangeMacro
+  if isFunctionLiteral(rhs) then in(Expr.betaReduce(rhs))
```

Comment:
Nice, in the old scala2 code there was a differentiation for literals as well that I didn't manage to port to scala 3